File: graph/adjacency-list.go
package main

import (
	"fmt"
	"slices"
)

type Graph struct {
	adjacencyList map[string][]string
}

func (g *Graph) addVertex(vertex string) bool {
	if _, ok := g.adjacencyList[vertex]; !ok {
		g.adjacencyList[vertex] = []string{}
		return true
	}

	return false
}

func (g *Graph) addEdge(vertex1 string, vertex2 string) {
	if _, ok := g.adjacencyList[vertex1]; ok {
		if idx := slices.Index(g.adjacencyList[vertex1], vertex2); idx == -1 {
			g.adjacencyList[vertex1] = append(g.adjacencyList[vertex1], vertex2)
		}
	}

	if _, ok := g.adjacencyList[vertex2]; ok {
		if idx := slices.Index(g.adjacencyList[vertex2], vertex1); idx == -1 {
			g.adjacencyList[vertex2] = append(g.adjacencyList[vertex2], vertex1)
		}
	}
}

func (g *Graph) removeEdge(vertex1 string, vertex2 string) {
    if edges, ok := g.adjacencyList[vertex1]; ok {
        if idx := slices.Index(edges, vertex2); idx != -1 {
            g.adjacencyList[vertex1] = append(edges[:idx], edges[idx+1:]...)
        }
    }

    if edges, ok := g.adjacencyList[vertex2]; ok {
        if idx := slices.Index(edges, vertex1); idx != -1 {
            g.adjacencyList[vertex2] = append(edges[:idx], edges[idx+1:]...)
        }
    }
}

func (g *Graph) removeVertex(vertex string) {
	defer delete(g.adjacencyList, vertex)
	if edges, ok := g.adjacencyList[vertex]; ok {
		for _, edge := range edges {
			g.removeEdge(vertex, edge)
		}
	}
}

func (g *Graph) dfsTraverseRecursive() []string {
	var start string
	traversedSlice := []string{}
	visitedVertexes := map[string]interface{}{}
	for key := range g.adjacencyList {
		start = key
		break;
	}

	g.dfsTraverseHelper(start, &traversedSlice, &visitedVertexes)

	return  traversedSlice
}

func (g *Graph) dfsTraverseHelper(vertex string, traversedSlice *[]string, visitedVertexes *map[string]interface{}) {
	edges, ok := g.adjacencyList[vertex]
	if !ok {
		return
	}

	*traversedSlice = append(*traversedSlice, vertex)
	(*visitedVertexes)[vertex] = true;

	for _, val := range edges {
		if _, ok := (*visitedVertexes)[val]; !ok {
			g.dfsTraverseHelper(val, traversedSlice, visitedVertexes)
		}
	}
}

func (g *Graph) dfsIterative() []string {
	var start string
		for key := range g.adjacencyList {
			start = key
			break;
		}

	stack := []string{start}	
	vertices := []string{}
	visited := map[string]bool{}
	visited[start] = true

	var currVertex string
	for (len(stack) != 0) {
		currVertex = stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		vertices = append(vertices, currVertex)

		for _, edge := range g.adjacencyList[currVertex] {
			if _, ok := visited[edge]; !ok {
				stack = append(stack, edge)
				visited[edge] = true
			}
		}
	}
	
	return vertices
}

func (g *Graph) bfs(start string) []string {
	queue := []string{start}
	result := []string{}
	visited := map[string]bool{}
	visited[start] = true
	var currVertex string

	for (len(queue) != 0) {
		currVertex = queue[len(queue)-1]
		queue = queue[:len(queue)-1]
		result = append(result, currVertex)

		if edges, ok := g.adjacencyList[currVertex]; ok {
			for _, edge := range edges {
				if _, ok := visited[edge]; !ok {
					visited[edge] = true;
					queue = append(queue, edge)
				}
			}
		}
	}

	return result;
}


func main() {
	graph := Graph{
		adjacencyList: map[string][]string{},
	}

	graph.addVertex("A");
	graph.addVertex("B");
	graph.addVertex("C");
	graph.addVertex("D");
	graph.addVertex("E");
	graph.addVertex("F");

	graph.addEdge("A", "B");
	graph.addEdge("A", "C");
	graph.addEdge("B", "D");
	graph.addEdge("C", "E");
	graph.addEdge("D", "E");
	graph.addEdge("D", "F");
	graph.addEdge("E", "F");
	fmt.Println(graph.dfsTraverseRecursive())
	fmt.Println(graph.dfsIterative())
	fmt.Println(graph.bfs("A"))
}

File: initial/graph/adjacency-list.js
class Graph {
	constructor() {
		this.adjacencyList = {};
	}

	/**
	 * add new vertex(node)
	 *
	 * @param {string | number} vertex
	 */
	addVertex(vertex) {
		if (this.adjacencyList[vertex]) return;

		this.adjacencyList[vertex] = [];
	}

	/**
	 * add edge(connection between two vertexes(nodes))
	 *
	 * @param {string | number} vertex1
	 * @param {string | number} vertex2
	 */
	addEdge(vertex1, vertex2) {
		const listForVertex1 = this.adjacencyList[vertex1];
		const listForVertex2 = this.adjacencyList[vertex2];

		if (listForVertex1?.indexOf(vertex2) === -1) {
			listForVertex1.push(vertex2);
		}

		if (listForVertex2?.indexOf(vertex1) === -1) {
			listForVertex2.push(vertex1);
		}
	}

	/**
	 * remove edge
	 *
	 * @param {string | number} vertex1
	 * @param {string | number} vertex2
	 */
	removeEdge(vertex1, vertex2) {
		const listForVertex1 = this.adjacencyList[vertex1];
		const listForVertex2 = this.adjacencyList[vertex2];

		const idxOfV2InV1 = listForVertex1?.indexOf(vertex2);
		const idxOfV1InV2 = listForVertex2?.indexOf(vertex1);

		if (idxOfV2InV1 >= 0) {
			listForVertex1.splice(idxOfV2InV1, 1);
		}

		if (idxOfV1InV2 >= 0) {
			listForVertex2.splice(idxOfV1InV2, 1);
		}
	}

	/**
	 * remove vertex(node)
	 *
	 * @param {string | number} vertex
	 */
	removeVertex(vertex) {
		const vertexEdges = this.adjacencyList[vertex];

		while (vertexEdges.length) {
			this.removeEdge(vertexEdges[0], vertex);
		}

		delete this.adjacencyList[vertex];
	}

	/**
	 * depth first tree traversal recursive
	 *
	 * @param {string | number} start
	 *
	 * @returns {string[]}
	 */
	dfsTraverseRecursive(start) {
		if (!start) start = Object.keys(this.adjacencyList)[0];

		const visited = new Set();
		const resArr = [];

		const traverse = (vertex) => {
			if (!vertex) return;
			visited.add(vertex);
			resArr.push(vertex);

			const edges = this.adjacencyList[vertex];

			for (const edge of edges) {
				if (!visited.has(edge)) traverse(edge);
			}
		};

		traverse(start);

		return resArr;
	}

	/**
	 * depth first tree traversal iterative
	 *
	 * @param {string | number} start
	 *
	 * @returns {string[]}
	 */
	dfsTraverseIterative(start) {
		if (!start) start = Object.keys(this.adjacencyList)[0];

		const vertices = [];
		const visited = new Set([start]);
		let stack = [start];

		let currVertex;
		while (stack.length) {
			currVertex = stack.pop();
			vertices.push(currVertex);

			this.adjacencyList[currVertex].forEach((edge) => {
				if (!visited.has(edge)) {
					visited.add(edge);
					stack.push(edge);
				}
			});
		}

		return vertices;
	}

	/**
	 * bredth first tree traversal iterative
	 *
	 * @param {string | number} start
	 *
	 * @returns {string[]}
	 */
	bfs(start) {
		const queue = [start];
		const visited = new Set(start);
		const result = [];

		let currVertex;
		while (queue.length) {
			currVertex = queue.pop();
			result.push(currVertex);

			this.adjacencyList[currVertex]?.forEach((edge) => {
				if (!visited.has(edge)) {
					visited.add(edge);
					queue.push(edge);
				}
			});
		}

		return result;
	}
}

const graph = new Graph();
graph.addVertex("A");
graph.addVertex("B");
graph.addVertex("C");
graph.addVertex("D");
graph.addVertex("E");
graph.addVertex("F");

graph.addEdge("A", "B");
graph.addEdge("A", "C");
graph.addEdge("B", "D");
graph.addEdge("C", "E");
graph.addEdge("D", "E");
graph.addEdge("D", "F");
graph.addEdge("E", "F");

// console.log(graph.dfsTraverseRecursive());
// console.log(graph.dfsTraverseIterative());
console.log(graph.bfs("A"));
